dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java

```diff
@@ -266,42 +221,112 @@ private boolean patternMatches(CachedVanityUrl cachedVanityUrl, String uri) {
     }
 
     /**
-     * Search CachedVanity Url checking if the Uri is in the Host Id and language Id cache
+     * Search CachedVanity Url checking if the Uri is in the Site Id and language Id cache
      *
      * @param uri The current uri
-     * @param hostId the current host Id
+     * @param siteId the current site Id
      * @param languageId the current language Id
      * @return a CachedVanityUrl object
      */
-    private CachedVanityUrl searchLiveCachedVanityUrlByHostAndLanguage(final String uri,
-            final String hostId, final long languageId) {
+    private CachedVanityUrl searchLiveCachedVanityUrlBySiteAndLanguage(final String uri,
+            final String siteId, final long languageId) {
+
         CachedVanityUrl result = null;
 
-        //Get the list of host cached Vanities URLs
-        Set<CachedVanityUrl> cachedVanityUrls = VanityUrlServices.getInstance()
-                .getVanityUrlByHostAndLanguage(hostId, languageId);
+        //Get the list of site cached Vanities URLs
+        Set<CachedVanityUrl> cachedVanityUrls = vanityUrlServices
+                .getVanityUrlBySiteAndLanguage(siteId, languageId);
 
         if (cachedVanityUrls.isEmpty()) {
-            //Initialize the Cached Vanity URL cache if is null
-            VanityUrlServices.getInstance().initializeVanityUrlCache(hostId, languageId);
-            cachedVanityUrls = VanityUrlServices.getInstance()
-                    .getVanityUrlByHostAndLanguage(hostId, languageId);
+
+            synchronized (VanityUrlAPIImpl.class) {
+
+                cachedVanityUrls = vanityUrlServices
+                        .getVanityUrlBySiteAndLanguage(siteId, languageId);
+
+                if (cachedVanityUrls.isEmpty()) {
+
+                    //Initialize the Cached Vanity URL cache if is null
+                    vanityUrlServices.initializeVanityUrlCache(siteId, languageId);
+
+                    //Get the list of site cached Vanities URLs
+                    cachedVanityUrls = vanityUrlServices
+                            .getVanityUrlBySiteAndLanguage(siteId, languageId);
+                }
+            }
         }
 
-        //Validates if onw of the host cachedVanityUrls matches the uri
+        //Validates if onw of the site cachedVanityUrls matches the uri
         for (CachedVanityUrl vanity : cachedVanityUrls) {
             if (patternMatches(vanity, uri)) {
                 result = vanity;
                 break;
             }
         }
+
+        /*
+        At this point we already found and saved in cache the VanityURL that matches with the given
+        URL but adding a new VanityURL to the cache for the completed requested URL
+        and not just the regex used can save us time.
+         */
+        if (null != result) {
+            if (!result.getUrl().equals(uri)) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)